bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java

```diff
@@ -1898,302 +1788,81 @@ void handleBookieFailure(final Map<Integer, BookieSocketAddress> failedBookies)
-                }
-                return false;
-            }
+    void ensembleChangeLoop(List<BookieSocketAddress> origEnsemble, Map<Integer, BookieSocketAddress> failedBookies) {
+        // when the ensemble changes are too frequent, close handle
+        if (numEnsembleChanges.incrementAndGet() > clientCtx.getConf().maxAllowedEnsembleChanges) {
             if (LOG.isDebugEnabled()) {
-                LOG.info("[EnsembleChange-L{}-{}] : resolved ledger metadata conflict and writing to zookeeper,"
-                        + " local meta data is \n {} \n, zk meta data is \n {}.",
-                        ledgerId, ensembleChangeIdx, metadata, newMeta);
+                LOG.debug("Ledger {} reaches max allowed ensemble change number {}",
+                          ledgerId, clientCtx.getConf().maxAllowedEnsembleChanges);
             }
-            // update znode version
-            metadata.setVersion(newMeta.getVersion());
-            // merge ensemble infos from new meta except last ensemble
-            // since they might be modified by recovery tool.
-            metadata.mergeEnsembles(newMeta.getEnsembles());
-            writeLedgerConfig(new ChangeEnsembleCb(ensembleInfo, curBlockAddCompletions,
-                    ensembleChangeIdx));
-            return true;
+            handleUnrecoverableErrorDuringAdd(WriteException);
+            return;
         }
 
-        @Override
-        public String toString() {
-            return String.format("ReReadLedgerMetadata(%d)", ledgerId);
-        }
+        new MetadataUpdateLoop(
+                clientCtx.getLedgerManager(), getId(),
+                this::getLedgerMetadata,
+                (metadata) -> {
+                    List<BookieSocketAddress> lastEnsemble = metadata.getLastEnsembleValue();
+                    boolean failedBookieInEnsemble = failedBookies.entrySet().stream()
+                        .anyMatch((e) -> lastEnsemble.get(e.getKey()).equals(e.getValue()));
+                    return !metadata.isClosed() && !metadata.isInRecovery() && failedBookieInEnsemble;
+                },
+                (metadata) -> {
+                    List<BookieSocketAddress> newEnsemble = EnsembleUtils.replaceBookiesInEnsemble(
+                            clientCtx.getBookieWatcher(), getId(), metadata, getCurrentEnsemble(), failedBookies);
+                    Long lastEnsembleKey = metadata.getLastEnsembleKey();
+                    LedgerMetadataBuilder builder = LedgerMetadataBuilder.from(metadata);
+                    long newEnsembleStartEntry = getLastAddConfirmed() + 1;
+                    checkState(lastEnsembleKey <= newEnsembleStartEntry,
+                               "New ensemble must either replace the last ensemble, or add a new one");
+                    if (lastEnsembleKey.equals(newEnsembleStartEntry)) {
+                        return builder.replaceEnsembleEntry(newEnsembleStartEntry, newEnsemble).build();
```

Comment:
Hmm; so the above checkSate should be < instead of <=??

Reply:
No, <, if we fail to write at all to an ensemble, it can be replaced.